diff --git a/lib/xmlparse.c b/lib/xmlparse.c
--- a/lib/xmlparse.c
+++ b/lib/xmlparse.c
@@ -608,7 +608,9 @@

 static XML_Char *copyString(const XML_Char *s, XML_Parser parser);

+#ifndef MOZILLA_CLIENT /* we already set a salt through XML_SetHashSalt */
 static unsigned long generate_hash_secret_salt(XML_Parser parser);
+#endif
 static XML_Bool startParsing(XML_Parser parser);

 static XML_Parser parserCreate(const XML_Char *encodingName,
@@ -1043,6 +1045,7 @@
        ASCII_s,     ASCII_p,     ASCII_a,      ASCII_c,      ASCII_e,
        '\0'};

+#ifndef MOZILLA_CLIENT /* we already set a salt through XML_SetHashSalt */
 /* To avoid warnings about unused functions: */
 #if ! defined(HAVE_ARC4RANDOM_BUF) && ! defined(HAVE_ARC4RANDOM)

@@ -1250,6 +1253,7 @@
   }
 #endif
 }
+#endif /* MOZILLA_CLIENT */

 static unsigned long
 get_hash_secret_salt(XML_Parser parser) {
@@ -1326,9 +1330,11 @@

 static XML_Bool /* only valid for root parser */
 startParsing(XML_Parser parser) {
+#ifndef MOZILLA_CLIENT /* we already set a salt through XML_SetHashSalt */
   /* hash functions must be initialized before setContext() is called */
   if (parser->m_hash_secret_salt == 0)
     parser->m_hash_secret_salt = generate_hash_secret_salt(parser);
+#endif
   if (parser->m_ns) {
     /* implicit context only set for root parser, since child
        parsers (i.e. external entity parsers) will inherit it
