diff --git a/lib/xmlparse.c b/lib/xmlparse.c
--- a/lib/xmlparse.c
+++ b/lib/xmlparse.c
@@ -2767,7 +2767,11 @@
   if (parser->m_eventPtr)
     return (XML_Index)(parser->m_parseEndByteIndex
                        - (parser->m_parseEndPtr - parser->m_eventPtr));
+#ifdef MOZILLA_CLIENT /* fix XML_GetCurrentByteIndex */
+  return parser->m_parseEndByteIndex;
+#else
   return -1;
+#endif
 }

 int XMLCALL
