diff --git a/lib/xmlparse.c b/lib/xmlparse.c
--- a/lib/xmlparse.c
+++ b/lib/xmlparse.c
@@ -3492,9 +3492,14 @@
       } else if (! entity) {
         if (parser->m_skippedEntityHandler)
           parser->m_skippedEntityHandler(parser->m_handlerArg, name, 0);
+#ifdef MOZILLA_CLIENT
+/* Bug 35984 - Undeclared entities are ignored when external DTD not found */
+        return XML_ERROR_UNDEFINED_ENTITY;
+#else
         else if (parser->m_defaultHandler)
           reportDefault(parser, enc, s, next);
         break;
+#endif
       }
       if (entity->open)
         return XML_ERROR_RECURSIVE_ENTITY_REF;
@@ -6790,7 +6795,12 @@
         if ((pool == &parser->m_tempPool) && parser->m_defaultHandler)
           reportDefault(parser, enc, ptr, next);
         */
+#ifdef MOZILLA_CLIENT
+/* Bug 35984 - Undeclared entities are ignored when external DTD not found */
+        return XML_ERROR_UNDEFINED_ENTITY;
+#else
         break;
+#endif
       }
       if (entity->open) {
         if (enc == parser->m_encoding) {
